### PR TITLE
feat(api): add promote workflow endpoint

### PR DESCRIPTION
**Pull request by ChmaraX**
### What changed? Why was the change needed?
* added SyncToEnvironment usecase
* added promote workflow endpoint
* added basic e2e tests to cover the endpoint
* fixed few bugs in other places linked to workflow CRUD

### CODE REVIEW

File: apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts [hunk 1]
```diff
@@ -34,6 +34,7 @@ export class GetWorkflowUseCase {
       GetPreferencesCommand.create({
         environmentId: command.user.environmentId,
         organizationId: command.user.organizationId,
+        templateId: workflowEntity._id,
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts:36]: We should provide `templateId` when retrieving specific workflow, otherwise it would return any preference from the org and environment I guess.
(existing e2e test was failing due to this)

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    const upsertedWorkflow = await this.upsertWorkflowUseCase.execute(
+      UpsertWorkflowCommand.create({
+        user: { ...command.user, environmentId: command.targetEnvironmentId },
+        identifierOrInternalId: existingWorkflow?._id,
+        workflowDto,
+      })
+    );
+
+    return upsertedWorkflow;
+  }
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts]: With some modification, this usecase could in theory be reused in Bridge sync, since the logic is very similar. 
This usecase would need to accept array of workflows instead of workflowId.

File: apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts [hunk 3]
```diff
@@ -123,15 +123,13 @@ export class UpsertWorkflowUseCase {
     command: UpsertWorkflowCommand,
     workflow: NotificationTemplateEntity
   ): Promise<GetPreferencesResponseDto | undefined> {
-    if (!command.workflowDto.preferences?.user) {
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts:128]: Removed this because we want to always execute `upsertPreferences`, but in case it gets empty preferences, it will insert default ones.

File: apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts [hunk 4]
```diff
@@ -123,15 +123,13 @@ export class UpsertWorkflowUseCase {
...
-      return undefined;
-    }
-    await this.upsertPreferences(workflow, command);
+    await this.upsertUserWorkflowPreferences(workflow, command);
+    await this.upsertWorkflowPreferences(workflow, command);
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts:130]: Added this to always insert both user and workflow preferences, in case they're not provided, we create default ones.

File: apps/api/src/app/workflows-v2/workflow.controller.e2e.ts [hunk 5]
```diff
@@ -193,6 +193,116 @@ describe('Workflow Controller E2E API Testing', () => {
       expect(idsDeduplicated.size).to.be.equal(10);
     });
   });
+
+  describe('Promote Workflow Permutations', () => {
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/workflow.controller.e2e.ts:228]: Given the fact that worfklow creation and promotion in general is a production critical part, I think it would make sense to cover `upsert-workflow.usecase.ts` with usecase specific tests to check if all entities are being created and updated correctly, by explicitly checking repositories.

These e2e tests serves more like a smoke tests, but they can't verify the functionality properly here.

I will create another task for this and add this in separate PR.

File: apps/api/src/app/workflows-v2/workflow.controller.ts [hunk 6]
```diff
@@ -63,6 +67,23 @@ export class WorkflowController {
...
+  @UseGuards(UserAuthGuard)
+  async promote(
+    @UserSession() user: UserSessionData,
+    @Param('workflowId') workflowId: string,
+    @Body() promoteWorkflowDto: PromoteWorkflowDto
+  ): Promise<WorkflowResponseDto> {
+    return this.syncToEnvironmentUseCase.execute(
+      SyncToEnvironmentCommand.create({
+        identifierOrInternalId: workflowId,
+        targetEnvironmentId: promoteWorkflowDto.targetEnvironmentId,
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/workflow.controller.ts:87]: I could wrap this usecase into `PromoteWorkflowUsecase` where I would retrieve environment with name "Production" to avoid the need for user to provide `targetEnvironmentId`. But I think this is a bit more flexible and we don't need additional usecase.

File: libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts [hunk 7]
```diff
@@ -61,12 +61,12 @@ export class UpsertPreferences {
...
-  private async upsert(
+  public async upsert(
     command: UpsertPreferencesCommand,
   ): Promise<PreferencesEntity> {
     const foundId = await this.getPreferencesId(command);
 
-    if (command.preferences === null) {
+    if (command.preferences === null && foundId) {
       return this.deletePreferences(command, foundId);
     }
```
Comment (ChmaraX) [inline libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: `this.deletePreferences` needs `foundId` for deletion. If the `foundId` is undefined, it throws and error.

File: packages/shared/src/dto/workflows/workflow-commons-fields.ts [hunk 8]
```diff
@@ -80,4 +80,5 @@ export type PreferencesResponseDto = {
 
 export type PreferencesRequestDto = {
   user: WorkflowPreferences | null;
+  workflow?: WorkflowPreferences | null;
```
Comment (ChmaraX) [inline packages/shared/src/dto/workflows/workflow-commons-fields.ts:81]: In case of promotion, we want to clone also workflow preferences from dev -> prod.

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    private getWorkflowByIdsUseCase: GetWorkflowByIdsUseCase,
+    private preferencesRepository: PreferencesRepository,
+    private upsertWorkflowUseCase: UpsertWorkflowUseCase
+  ) {}
+
+  async execute(command: SyncToEnvironmentCommand): Promise<WorkflowResponseDto> {
+    const workflowToClone = await this.getWorkflowToClone(command);
+    const preferencesToClone = await this.getWorkflowPreferences(workflowToClone._id, workflowToClone._environmentId);
+    const externalId = workflowToClone.triggers[0].identifier;
+    const existingWorkflow = await this.findExistingWorkflow(command, externalId);
```
Comment (LetItRock) [inline apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts]:
```suggestion
    const existingWorkflow = await this.findWorkflowInTargetEnvironment(command, externalId);
```

File: apps/api/src/app/workflows-v2/workflow.controller.e2e.ts [hunk 10]
```diff
@@ -193,6 +193,116 @@ describe('Workflow Controller E2E API Testing', () => {
...
+      expect(prodWorkflow.steps).to.have.lengthOf(devWorkflow.steps.length);
+      prodWorkflow.steps.forEach((prodStep, index) => {
+        const devStep = devWorkflow.steps[index];
+        expect(prodStep.stepUuid).to.not.equal(devStep.stepUuid, 'Step UUID should be different');
+        expect(prodStep.controlValues).to.deep.equal(devStep.controlValues, 'Step controlValues should match');
+        expect(prodStep.name).to.equal(devStep.name, 'Step name should match');
+        expect(prodStep.type).to.equal(devStep.type, 'Step type should match');
+      });
+    });
+
```
Comment (LetItRock) [inline apps/api/src/app/workflows-v2/workflow.controller.e2e.ts:270]: I guess we should not allow promotion in the same environment, right? if yes, let's add the test and validation

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+      /*
+       * If we are updating an existing workflow, we need to map the updated steps to the existing steps
+       * (!) 'existingWorkflowSteps' are from a different environment than 'steps' - the only thing that doesn't change
+       *  in steps across environments is the stepId (TODO)
+       */
+      ...(existingWorkflowSteps && {
+        _id:
+          existingWorkflowSteps.find((existingStep) => existingStep.stepId === step.stepId)?._templateId ??
+          step._templateId,
+      }),
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts:139]: This doesn't matter yet, because `stepId` is not kept the same between the same workflow in different environments, therefore steps are always deleted & recreated.

Once we make `stepId` (or other step property) immutable, this will work as expected

File: apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts [hunk 12]
```diff
@@ -207,7 +223,7 @@ export class UpsertWorkflowUseCase {
 
     return {
       id: existingWorkflow._id,
-      environmentId: user.environmentId,
+      environmentId: existingWorkflow._environmentId,
```
Comment (ChmaraX) [inline apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts:235]: This needs to be the environment where the workflow that we are updating is stored, not the user current environment.

**Review by SokratisVidros — APPROVED** [review]

File: apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts (hunk 1 above)
Comment (rifont) [reply apps/api/src/app/workflows-v2/usecases/get-workflow/get-workflow.usecase.ts:36]: Good catch!

File: apps/api/src/app/workflows-v2/workflow.controller.e2e.ts [hunk 13]
```diff
@@ -225,6 +225,130 @@ describe('Workflow Controller E2E API Testing', () => {
...
+    it('should throw an error if the workflow to promote is not found', async () => {
+      const res = await session.testAgent.put(`${v2Prefix}/workflows/123/promote`).send({ targetEnvironmentId: '123' });
+
+      expect(res.status).to.equal(404);
+      expect(res.body.message).to.equal('Workflow cannot be found');
+      expect(res.body.workflowId).to.equal('123');
+    });
+  });
+
```
Comment (rifont) [inline apps/api/src/app/workflows-v2/workflow.controller.e2e.ts:351]: Nice tests ✨

File: libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts (hunk 7 above)
Comment (rifont) [reply libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: You're spot on with the `foundId` part, I made the same change in my PR.

File: libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts [hunk 14]
```diff
@@ -61,12 +61,12 @@ export class UpsertPreferences {
     });
   }
 
-  private async upsert(
+  public async upsert(
```
Comment (rifont) [inline libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: ⚠ **issue:** Only the `execute` method should be `public` in use-cases. We should revert this change and use `execute` as needed by passing in the necessary command properties to control the use-case behaviour.‏
```suggestion
  private async upsert(
```

Comment (rifont) [reply libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: I see that other methods are `public` in the `UpsertPreferences` - I will address those in a follow-up PR, but we should avoid increasing the problem for now.

**Review by rifont — APPROVED** [review]
Very clean! Nice work 🎉

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts:1]: How would we feel about renaming `sync` to something else? I think using `sync` here might be confusing since we already use it to sync data from the bridge app (like workflows, preferences, etc.). 

This is similar but has its own differences, so I wonder if it would be better to avoid reusing the same term.

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+      workflow: preferences.find((pref) => pref.type === PreferencesTypeEnum.WORKFLOW_RESOURCE)?.preferences ?? null,
+    };
+  }
+
+  private async getWorkflowPreferences(workflowId: string, environmentId: string): Promise<PreferencesEntity[]> {
+    const workflowPreferences = await this.preferencesRepository.find({
+      _templateId: workflowId,
+      _environmentId: environmentId,
+      type: {
+        $in: [PreferencesTypeEnum.WORKFLOW_RESOURCE, PreferencesTypeEnum.USER_WORKFLOW],
```
Comment (djabarovgeorge) [inline apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts:164]: question because i am missing the context, do we allow to `:workflowId/promote` code-first workflows as well?

File: apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts (hunk 4 above)
Comment (djabarovgeorge) [reply apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts:130]: thought: agree on `user` preference default upsert but, i wonder if we need to create a default `workflow` preference for users that do not use code-first, feels a bit redundant and misleading data.

File: apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts (hunk 12 above)
Comment (djabarovgeorge) [reply apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts:235]: good point, besides the tests i would add a const with a comment to prevent future maintainers break it while refactoring the code. 
not a deal breaker just as a thought for your consideration.

File: apps/api/src/app/workflows-v2/workflow.controller.ts (hunk 6 above)
Comment (djabarovgeorge) [reply apps/api/src/app/workflows-v2/workflow.controller.ts:87]: i think we will have only Prod to promote to in the near-far future but as long as the API client knows the prod env i think this approach is good enough and as you said more flexible let's keep it for now.

File: libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts (hunk 7 above)
Comment (djabarovgeorge) [reply libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: great mind think alike 😉

**Review by djabarovgeorge — COMMENTED** [review]
amazing work, super clean!

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts (hunk 15 above)
Comment (ChmaraX) [reply apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts:164]: I think for now we agreed to **not** promote code-first workflows using dashboard UI, because they were some open questions and issues with that.

File: libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts (hunk 14 above)
Comment (ChmaraX) [reply libs/application-generic/src/usecases/upsert-preferences/upsert-preferences.usecase.ts]: Oh, I was using upsert directly at some point then I removed that implementation and forgot to put the method back to private.

File: apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts
Comment (ChmaraX) [reply apps/api/src/app/workflows-v2/usecases/sync-to-environment/sync-to-environment.usecase.ts:1]: I agree its not 100% clear what it does, however I was thinking maybe with some small adjustments we could use it also in in the place of bridge sync, then the naming would be alright I guess.

I'm was thinking about `promote.usecase` but thats less flexible or maybe something with `upsert` or `clone`.
